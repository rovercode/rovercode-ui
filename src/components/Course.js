import React from 'react';
import { hot } from 'react-hot-loader';
import {
  Accordion,
  AccordionSummary,
  AccordionDetails,
  Grid,
  Typography,
} from '@material-ui/core';
import { ExpandMore } from '@material-ui/icons';
import PropTypes from 'prop-types';

import LessonCard from './LessonCard';

const Course = ({ course, userTier, onLessonClick }) => (
  <Accordion defaultExpanded>
    <AccordionSummary expandIcon={<ExpandMore />}>
      <Typography>
        { course.name }
      </Typography>
    </AccordionSummary>
    <AccordionDetails>
      <Grid container direction="row" spacing={2}>
        {
          course.lessons.sort((l1, l2) => l1.sequence_number - l2.sequence_number)
            .map((lesson) => (
              <Grid item xs={12} md={6} lg={3} key={lesson.id}>
                <LessonCard lesson={lesson} userTier={userTier} onClick={onLessonClick} />
              </Grid>
            ))
        }
      </Grid>
    </AccordionDetails>
  </Accordion>
);

Course.defaultProps = {
  userTier: 1,
};

Course.propTypes = {
  course: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    lessons: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      reference: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
    })).isRequired,
  }).isRequired,
  userTier: PropTypes.number,
  onLessonClick: PropTypes.func.isRequired,
};

export default hot(module)(Course);
