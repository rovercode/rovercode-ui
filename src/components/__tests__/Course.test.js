import React from 'react';
import { shallow } from 'enzyme';

import Course from '../Course';

describe('The Course component', () => {
  const course = {
    id: 1,
    name: 'Test',
    description: 'This is a test',
    lessons: [{
      id: 1,
      reference: 'Lesson1',
      description: 'The second lesson',
      sequence_number: 2,
      active_bd: 1,
      active_bd_owned: false,
      state: {
        progress: 'IN_PROGRESS',
      },
    }, {
      id: 2,
      reference: 'Lesson2',
      description: 'The first lesson',
      sequence_number: 1,
      active_bd: 2,
      active_bd_owned: true,
      state: {
        progress: 'AVAILABLE',
      },
    }],
  };
  const onLessonClick = jest.fn();

  test('renders on the page with no errors', () => {
    const wrapper = shallow(<Course course={course} onLessonClick={onLessonClick} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('sorts the lessons', () => {
    const wrapper = shallow(<Course course={course} onLessonClick={onLessonClick} />);

    expect(wrapper.find('LessonCard').first().prop('lesson').id).toBe(2);
    expect(wrapper.find('LessonCard').last().prop('lesson').id).toBe(1);
  });

  test('handles accordion expansion', () => {
    const wrapper = shallow(<Course course={course} onLessonClick={onLessonClick} />);

    expect(wrapper.find('WithStyles(ForwardRef(Typography))').at(1).text()).toBe('Hide Course');
    wrapper.find('WithStyles(ForwardRef(Accordion))').props().onChange();
    expect(wrapper.find('WithStyles(ForwardRef(Typography))').at(1).text()).toBe('Show Course');
  });
});
