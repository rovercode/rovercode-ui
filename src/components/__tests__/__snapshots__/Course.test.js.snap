// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Course component renders on the page with no errors 1`] = `
<WithStyles(ForwardRef(Accordion))
  defaultExpanded={true}
>
  <WithStyles(ForwardRef(AccordionSummary))
    expandIcon={<Memo(ExpandMoreIcon) />}
  >
    <WithStyles(ForwardRef(Typography))>
      Test
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(AccordionSummary))>
  <WithStyles(ForwardRef(AccordionDetails))>
    <WithStyles(ForwardRef(Grid))
      container={true}
      direction="row"
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        key="2"
        lg={3}
        md={6}
        xs={12}
      >
        <LessonCard
          lesson={
            Object {
              "active_bd": 2,
              "active_bd_owned": true,
              "description": "The first lesson",
              "id": 2,
              "reference": "Lesson2",
              "sequence_number": 1,
              "state": Object {
                "progress": "AVAILABLE",
              },
            }
          }
          onClick={[MockFunction]}
          userTier={1}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        key="1"
        lg={3}
        md={6}
        xs={12}
      >
        <LessonCard
          lesson={
            Object {
              "active_bd": 1,
              "active_bd_owned": false,
              "description": "The second lesson",
              "id": 1,
              "reference": "Lesson1",
              "sequence_number": 2,
              "state": Object {
                "progress": "IN_PROGRESS",
              },
            }
          }
          onClick={[MockFunction]}
          userTier={1}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(AccordionDetails))>
</WithStyles(ForwardRef(Accordion))>
`;
